### PR TITLE
Ruby Exercises

### PR DESCRIPTION
**Pull request by architag**

### CODE REVIEW

File: Ruby/exercise_4_palindrome.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+class String
+  def palindrome?
+    if self == self.reverse
```
Comment (neeraj33negi) [inline Ruby/exercise_4_palindrome.rb:3]: no need to use if - else, simply return `self == self.reverse`

File: Ruby/exercise_4_palindrome.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    else
+      return false
+    end
+  end
+end
+
+empty_string = true
+while empty_string
+  puts "Please provide an input"
+  input = gets
```
Comment (neeraj33negi) [inline Ruby/exercise_4_palindrome.rb:14]: use have to use command line arguments for input

Comment (neeraj33negi) [reply Ruby/exercise_4_palindrome.rb:14]: and if there is no input from args, then ask for input

File: Ruby/exercise_5_customer_account_balance.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
+class Account
+  @@account_no = 0;
```
Comment (neeraj33negi) [inline Ruby/exercise_5_customer_account_balance.rb:2]: this is fine for now, but remember that class variables are not thread safe and can lead to race conditions.

File: Ruby/exercise_5_customer_account_balance.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+input = gets
+data = Array.new()
+details, amt = input.split(' transfer:')
+accounts = details.scan(/[A-Za-z]+:[0-9]+|"[A-Za-z]+ [A-Za-z]+":[0-9]+/)
+accounts.each do |var|
+  var.gsub!("\"", "")
+  data << var.split(':')
+end
+
+acc_1 = Account.new(data[0][0], Integer(data[0][1]))
```
Comment (neeraj33negi) [inline Ruby/exercise_5_customer_account_balance.rb:30]: not a strong requirement for this exercise but can we make this transaction an atomic operation?

File: Ruby/exercise_6_vehicle_subclass.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    "Name: #{@name} \nPrice: #{@price}"
+  end
+end
+
+class Bike < Vehicle
+  def initialize(name, price, dealer, percent_price_increase)
+    super(name, price)
+    @dealer = dealer
+    @percent_price_increase = percent_price_increase
+  end
```
Comment (neeraj33negi) [inline Ruby/exercise_6_vehicle_subclass.rb:16]: spaces between two method definitions

File: Ruby/exercise_6_vehicle_subclass.rb [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  def price_increase
+    @price *= 1 + (@percent_price_increase/100)
+  end
```
Comment (neeraj33negi) [inline Ruby/exercise_6_vehicle_subclass.rb:19]: space

File: Ruby/exercise_10_hash_inject.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+arr = Array.new()
+
+input.gsub!(/\[|\]/, "")
+input.split(',').each do |var|
+  var.gsub!(/"|'/, "")
+  arr << var
+end
+
+output = hashInject(arr.createHash())
+puts output
```
[the remark this answers is not in the document]
Comment (architag) [reply Ruby/exercise_10_hash_inject.rb:32]: The input format mentioned in the question is like "['abc','def',1234,234]". 
Output shows {"odd"=>[["abc", "def", "234"]], "even"=>[["1234"]]} with this.

File: Ruby/exercise_20_interest_difference.rb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  def initialize
+    yield self
+  end
+
+  def calc_difference
+    rate = 0.1
+    comp = (@principal*(1+rate)**@time)
+    si = (@principal*(1+rate*@time))
+    diff = (comp - si).round(2)
```
Comment (neeraj33negi) [inline Ruby/exercise_20_interest_difference.rb:12]: dont have to assign to a new variable, just return the diff

File: Ruby/exercise_21_factorial.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+def factorial(n)
+  if n==0
+    return 1
+  end
+  (1..n).inject(:*)
+end
+
+begin
+  input = Integer(ARGV[0])
+  raise "Negative number entered" if input < 0
```
Comment (neeraj33negi) [inline Ruby/exercise_21_factorial.rb:10]: This returns RunTimeError, it should be an ArgumentError.

File: Ruby/exercise_22_name_raise.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  end
+end
+
+puts "Please provide an input"
+input = gets.chomp
+input.gsub!("\"", "")
+
+begin
+  firstname, lastname = input.split(" ")
+  raise "Firstname cannot be blank" if !firstname
```
Comment (neeraj33negi) [inline Ruby/exercise_22_name_raise.rb:17]: Please create custom exception class for this exercise.

File: Ruby/exercise_24_reverse_iterate.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+class Array
+  def reverse_iterate
+    i=self.length-1
```
Comment (neeraj33negi) [inline Ruby/exercise_24_reverse_iterate.rb:3]: leave space around operators `i = self.length - 1`

File: Ruby/exercise_25_sales_tax.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
+sum = 0
```
Comment (neeraj33negi) [inline Ruby/exercise_25_sales_tax.rb:1]: Please try to implement this in a OOP pattern, identify if there could be a class and what could be methods

**Review by neeraj33negi — APPROVED** [review]

**Pull request merged**